File: 0041_First_Missing_Positive/First_Missing_Positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();        // Taking the size of the input

        int i = 0;                  // Counter
        while (i < n) {
            if (nums[i] > 0 && nums[i] < n + 1 && nums[i] != nums[nums[i] - 1]) { // Checking the range. Cause out of this range doesnt matter. Also checking if the correct value is in its correct position or it will loop infinitely.
                swap(nums[i], nums[nums[i] - 1]); // Swap
            } else {
                i++;
            }
        }

        for (i = 0; i < n; i++) {
            if (nums[i] != i + 1)
                return i + 1;
        }
        return n + 1;
    }
};

File: 0041_First_Missing_Positive/First_Missing_Positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)       # Getting length of List
        i = 0               # Counter

        while i < n:
            correctindex = nums[i] - 1      # Unlike C++, Pyhton is not that much intelligent to calculate inside the List 3rd brackets
            if 0 < nums[i] < n+1 and nums[i] != nums[correctindex]:
                nums[i], nums[correctindex] = nums[correctindex] , nums[i]      # Swap
            else:
                i += 1      # Increment
            
        for i in range(n):
            if nums[i] != i+1:      # At which element we could not find the next index same that one is the missing
                return i + 1        # Returning 
        
        return n + 1        # If 1...n all are available then n+1 is the missing one